### ISSUE TITLE
[META] Write documentation

### ISSUE THREAD

**Issue by jenlampton** [backdrop-ops/docs.backdropcms.org#3]
In addition to change records, the API site will hold content for documentation on specific areas of Backdrop.  There's [a rough equivalent](https://api.drupal.org/api/drupal/groups) to this on api.drupal.org that uses organic groups.

For Backdrop's API site we can probably get away with a simple Documentation content type with a body field (and perhaps taxonomy for the major version of Backdrop).
- [x]  Developer Documentation
  - [x] Converting Drupal code to Backdrop code
    - [x]  Change Records
    - [x]  Converting modules
    - [x]  Converting themes
    - [ ]  Creating new layouts (based on Drupal themes)
    - [ ]  Creating new layouts (based on panels layouts)
  - [x]  Contribute to Backdrop CMS
    - [x]  GitHub workflow
    - [x]  Contrib application process
  - [x]  Developing modules
    - [x]  Creating modules
    - [x]  Understanding the hook system
    - [ ]  Multilingual support
    - [ ]  Working with Views
    - [ ]  Creating forms
  - [x]  Developing layouts
    - [x]  Creating layouts
  - [x]  Developing themes
    - [x]  Creating themes
    - [ ]  Admin themes
    - [ ]  Template overrides
  - [ ]  Advanced API documentation
    - [ ]  Batch API
    - [ ]  Cache API
    - [x]  Database API
    - [ ]  Entity API
    - [ ]  Field API
    - [ ]  File API
    - [x]  Form API
    - [ ]  JavaScript API
    - [ ]  Layout API
    - [ ]  Localization API
    - [ ]  RDF Mapping API
    - [ ]  Render Arrays
    - [ ]  Schema API
    - [ ]  Theme API
    - [ ]  Views API
    - [ ]  XML-RPC
  - [x]  Coding standards
    - [x]  PHP Coding Standards
    - [x]  CSS Coding Standards
    - [x]  JS Coding Standards
  - [x]  Documentation Standards
  - [x]  Security
    - [x]  Writing secure code
    - [ ]  Security policy
  - [ ]  Accessibility best practices
- [x]  Backdrop API
  - [x]  Topics
  - [x]  Functions
  - [x]  Classes
  - [x]  Files
  - [x]  Globals
- [x]  Glossary

**Comment by jenlampton** [comment]
I just created a book here called Developer Documentation (https://api.backdropcms.org/documentation) and would love some help fleshing out these first few pages.  We can copy similar pages off drupal.org and change content where necessary.

@docwilmot let me know if you are interested in taking any of this on. I gave your account on the API site access, just in case :)

**Comment by BWPanda** [comment]
I'd like to recommend a few changes:

- Create a new link 'Standards' (or similar) and make 'PHP Coding Standards', 'CSS Coding Standards', 'JavaScript Coding Standards' & 'Code Documentation Standards' sub-links of it. The new 'Standards' link should also probably be made a sub-link of 'Developer Documentation'.

- Split up the 'Creating modules' & 'Creating themes' pages into smaller pages as sub-links of the main 'Creating *' page (e.g. move the 'More about module .info files' to a separate page under 'Creating modules'). This'll greatly help readability and comprehension.

**Comment by docwilmot** [comment]
> Create a new link 'Standards' (or similar) and make 'PHP Coding Standards', 'CSS Coding Standards', 'JavaScript Coding Standards' & 'Code Documentation Standards' sub-links of it. The new 'Standards' link should also probably be made a sub-link of 'Developer Documentation'.

Good plan

> This'll greatly help readability and comprehension.

Agreed. But I have issues with having too much depth in navigating docs, like Drupals docs do. So perhaps before we start chopping up, I hoped to come up with a 'table of contents' of the topics that should be in documetnation - priority topics first - and share that. Any suggestions as to what these priorities should be?
I also planned to try and understand Layout a bit more before I wrote it, and havent gotten there yet. Will also need sections on converting themes/modules/layouts from D7 to backdrop.

**Comment by jenlampton** [comment]
I also agree that there are too many pages in Drupal's docs, and would love to see a proposed outline before we do too much dividing.  I'd also prefer fewer pages that people can scan or search to find what they need, rather than lots and lots of little pages.

**Comment by docwilmot** [comment]
Here's a stab at a minimal outline for *developer* docs. (need to consider user help pages separately on backdropcms.org)

-  Developer Documentation
  -  Contribute to development
    -  GitHub workflow
    -  Contrib application process
  -  Developing modules
    -  Creating modules
    -  Converting modules from Drupal
    -  Understanding the hook system
    -  Multilingual support
    -  Working with Views
    -  Creating forms
  -  Developing layouts
    -  Creating layouts
    -  Converting themes to layouts
  -  Developing themes
    -  Creating themes
    -  Converting themes from Drupal
    -  Admin themes
    -  Template overrides
  -  Advanced API documentation
    -  Batch API
    -  Cache API
    -  Database API
    -  Entity API
    -  Field API
    -  File API
    -  Form API
    -  JavaScript API
    -  Layout API
    -  Localization API
    -  RDF Mapping API
    -  Render Arrays
    -  Schema API
    -  Theme API
    -  Views API
    -  XML-RPC
  -  Coding standards
    -  PHP Coding Standards
    -  CSS Coding Standards
    -  JS Coding Standards
  -  Documentation Standards
  -  Security
    -  Writing secure code
    -  Security policy
  -  Accessibility best practices
  -  Glossary
-  Change Records
-  Backdrop API
  -  Topics
  -  Functions
  -  Classes
  -  Files
  -  Globals

**Comment by cellear** [comment]
On behalf of outsiders, I think we should have an easy-to-find spot that explains how our particular ecosystem understands the different roles and tasks involved in constructing web software.  I'm always astonished at how Balkanized the web development world is - terms like "developer" and "user" mean different things to different communities.  (For example, I recently heard on a podcast somebody opine that they couldn't imagine somebody calling themselves a web developer that hadn't mastered at least 2 toolchains.  I've been building websites professionally for 20 years, and prior to a few months ago I had never even _heard_ the word "toolchain.")

The document I have in mind is very different from a glossary; it's more akin to an introduction or preface.  I'm willing to take a stab at it, if it sounds useful to anybody besides me!

**Comment by docwilmot** [comment]
Incidentally, the glossary is already done https://api.backdropcms.org/glossary

**Comment by docwilmot** [comment]
And I've been building websites amateurly for 12 years and never heard about toolchains till about 2 minutes ago. :)

**Comment by jenlampton** [comment]
@docwilmot wow, this outline is fantastic :)

How would you feel about separating "building new things" from "converting things"?

For the short term, people will want to be able to get at all the "converting things" documentation in one place, without having to wade through all the other pages.  In the long term, this section will also become less and less important, as we gain more Backdrop users who don't come from Drupal, and as there are fewer and fewer sites still running Drupal 7.

**Comment by docwilmot** [comment]
> How would you feel about separating "building new things" from "converting things"?

Makes a lot of sense.
Were [having a chat about the Form API pages](https://github.com/backdrop-ops/api.backdropcms.org/issues/6#issuecomment-71907734) just now. Views, FormAPI have had their own API pages in the past, but I'm not sure how those are built.  How do we include the API docs like those? 
Should we keep standalone API repos and link to those or just paste all that into a node or something else?

**Comment by quicksketch** [comment]
These outlines look great! I like @jenlampton's suggestion of separating out the upgrading docs as well. That will allow for us to do the same thing as needed for Backdrop 1 to Backdrop 2 docs.

**Comment by jenlampton** [comment]
@docwilmot I think webchick wrote all the Form API docs in a HTML editor and pasted that into a node for the drupal docs, it would be nice if we could do something similar in style (but perhaps easier to maintain?) If you think a separate git repo is the way to go we can certainly do that. We could do something fancy with a custom callback generating a page that's pulled from a repo...

For the Views API I think all the code docs should already be merged into the API site automatically, we'll need to take a look at what (if anything) is still relevant (were all the docs updated to views 3? I thought not...) and see how to incorporate those. 

I'm open to suggestions.

**Comment by quicksketch** [comment]
> I think webchick wrote all the Form API docs in a HTML editor and pasted that into a node for the drupal docs, it would be nice if we could do something similar in style

We got the Form API page up in https://github.com/backdrop-ops/api.backdropcms.org/issues/6.

**Comment by jenlampton** [comment]
Is it on github?

**Comment by cellear** [comment]
Isn't there a way to store content in a repo?  This is my pet issue, my single biggest complaint with Drupal.  If there's an easy "Drupal-style" way to do it, let's do it that way.  If not, I'll write it for Backdrop.  I desperately want it! (@jenlampton, I think you may have referenced "update hooks" as a way to do this, but I wasn't familiar with that technique.)

**Comment by quicksketch** [comment]
@cellear Are you suggesting that we'd have nodes (or some other data type) stored in Git? I don't think there's any such thing, because it would largely invalidate a huge number of Backdrop/Drupal concepts. Storing content in Git is directly at odds with database-driven content, making things like Views and Fields all but impossible. The only approach that would sort-of work would be replicating data from the Git repository *into* the database (as nodes or other data), but then you'd end up with a difficult situation how to reconcile the Drupal editing interface with the Git repository, or when something is updated in Git, you'd have to update the node.

All of this *could* be done, but it'd be significantly more complicated than I think we'd be interested in supporting.

**Comment by larsdesigns** [comment]
@cellear Have a look at Gollum: https://github.com/gollum/gollum

**Comment by cellear** [comment]
> Are you suggesting that we'd have nodes (or some other data type) stored in Git? 

@quicksketch: Yes, that's what I was hoping for :)  I'm envisioning something like the way Features works in Drupal.  It would pull data from "code" and create content - nodes, terms, users, etc. Once in the DB, this data would be fully managed the way we expect content to be.  Perhaps it could be reverted to the disk image, as Features allows, but if it can't it's not a deal-breaker - a one-time load gets me 80% of what I would be looking for. (Though I'm sure once I had that I'd come looking for some of that remaining 20%.)

Does Backdrop store things like Views and Content Types in the DB, or are they immediately written to CMI as JSON?  In my dreams, I'd love to be able to hit a switch and have Backdrop dump all of its content into a directory full of JSON files, which I could then load into another Backdrop instance.  I might try to write that, if it comes to it.  It's not impossible, is it?

@larsdesigns: I prefer to do my work in Backdrop, or possibly Drupal.  Switching to a whole other platform loses all that great stuff that brought me here in the first place!

**Comment by cellear** [comment]
I realize these Drupal projects have wrestled with this issue:

**Comment by cellear** [comment]
I realize these Drupal projects have wrestled with this problem area, with varying degrees of success:
- https://www.drupal.org/project/defaultcontent
- https://www.drupal.org/project/import
- https://www.drupal.org/project/migrate
- https://www.drupal.org/project/uuid_features
- https://www.drupal.org/project/deploy
- https://www.drupal.org/project/feeds
- https://www.drupal.org/project/node_import
- https://www.drupal.org/project/entity_pilot

Among others.  Not sure which of these hew closest to Backdrop's philosophy.  _All_ of them seem inferior to the import/export capabilities that come with Wordpress, but I haven't used those enough to know if whether they actually live up to their promise.

**Comment by quicksketch** [comment]
> Does Backdrop store things like Views and Content Types in the DB, or are they immediately written to CMI as JSON? In my dreams, I'd love to be able to hit a switch and have Backdrop dump all of its content into a directory full of JSON files, which I could then load into another Backdrop instance. I might try to write that, if it comes to it. It's not impossible, is it?

Backdrop stores all configuration directly as JSON files, with no database involved at all. Backdrop can already do full site export/imports to move everything (Views, Content Types, Fields, etc.) all at once from one site to another. There are limits though: we don't handle enabling/disabling modules or deletion of most things if they contain data (e.g. a field that has content in the database). The validation will stop you. But for 90% situations, everything already works this way. It is pretty much the #1 feature of Backdrop after all. :wink: 

That said, I think @larsdesigns has a good suggestion looking into gollum. Backdrop and Drupal both are *database-driven*, and even though configuration is stored in CMI now, *content* is still in the database and it's going to be an uphill battle making content store in a file structure so it could be in Git.

**Comment by docwilmot** [comment]
I must confess I dont get the 'store content in a repo' discussion, not quite sure what you guys are talking about, gollum, etc etc. But I was waiting for this to resolve to move on. Can we get a word on what next? Move on with putting things in Book module or wait?

**Comment by quicksketch** [comment]
My suggestion for "looking into gollum" was intended for @cellear. For now I'm pretty sure our only good option is to proceed with Book module. We were actually using Github pages for the handbook early on and that turned into a mess pretty quickly. There may be a Backdrop-based file solution at some point, but an adequate file-based solution for Drupal, I don't think we're likely to see one either. The two concepts of file-based management and database-stored management are opposed to each other.

@docwilmot if you need any help or permissions to start implementing the proposed outline, let @jenlampton or I know. I think the outline looks great and already the API site is shaping up nicely. If we do come up with a new mechanism later, we'll still need to have the content; so let's not hold up content on architecture.

**Comment by docwilmot** [comment]
Thanks. One advantage of having docs in a repository I suspect would probably be versioning. It would be great if we could think up some way of versioning developer docs as well, but based on core.
Just as the API docs have paths showing the core version (api.backdropcms.org/api/backdrop/.../1), would be good to have the same apply for docs too (api.backdropcms.org/creating-layouts/1).

**Comment by jenlampton** [comment]
Git is completely the wrong tool for versioning text -- code yes, english, no. @cellear have you ever tried it?  I tried to work on one of my Backdrop articles in Drupal watchdog that way - bad idea. 

If we need text versioning for our docs we could use revisions like d.o does.

**Comment by docwilmot** [comment]
I am going to put fiddling and porting modules aside for now and work on getting this and [this](https://github.com/backdrop-ops/backdropcms.org/issues/55) done. 

@jenlampton could you move your suggested outline to the first post and make it a checklist please?

I'm creating a folder of Google docs [here](https://drive.google.com/folderview?id=0B5vNMqFUUd-SflU4M25jb3VaTXZDVHJlUldubGtFcnJsb3J0RHBTM3BqT0FFTlhGT1N2alE&usp=sharing). That link gives edit permission as well.

Will need to be able to check them off as we go.

Thanks.

**Comment by jenlampton** [comment]
@docwilmot I added a checklist in the main issue, sorry for the delay on that!

**Comment by serundeputy** [comment]
I checked a bunch of stuff of the list in the OP.  Looks great! Everyone has been working hard look how far we've come!
